tests/Unit/ActionTesting.hpp

```diff
@@ -61,6 +61,238 @@ struct get_array_index;
+ * in a parallel executable, but instead must be advanced by calling the
+ * `set_phase` method of the `runner`. While testing the `TerminatePhase` action
+ * we are only interesting in checking whether the algorithm has set the
+ * `terminate` flag. This is done using the `ActionTesting::get_terminate()`
+ * function and is done for each distributed object (i.e. per component). The
+ * next action in the action list for the current phase is invoked by using the
+ * `ActionTesting::next_action()` function, which takes as arguments the runner
+ * by `not_null` and the array index of the distributed object to invoke the
+ * next action on.
+ *
+ * ### InitializeDataBox and action introspection
+ *
+ * Having covered the very basics of ATF let us look at the test for the
+ * `Actions::Goto` action. We will introduce the functionality
+ * `InitializeDataBox`,`force_next_action_to_be`, `get_next_action_index`, and
+ * `get_databox_tag`. The `Goto<the_label_class>` action changes the next action
+ * in the algorithm to be the `Actions::Label<the_label_class>`. For this test
+ * the Metavariables is:
+ *
+ * \snippet Test_Goto.cpp metavariables
+ *
+ * You can see that this time there are two test phases, `TestGoto` and
+ * `TestRepeatUntil`.
+ *
+ * The component for this case is quite a bit more complicated so we will go
+ * through it in detail. It is given by
+ *
+ * \snippet Test_Goto.cpp component
+ *
+ * Just as before there are `metavariables`, `chare_type`, and `array_index`
+ * type aliases. The `repeat_until_phase_action_list` is the list of iterable
+ * actions that are called during the `TestRepeatUntil`. The `Initialization`
+ * phase action list now has the action `ActionTesting::InitializeDataBox`,
+ * which takes simple tags and compute tags that will be added to the DataBox in
+ * the initial phase. How the values for the simple tags are set will be made
+ * clear below when we discuss the
+ * `ActionTesting::emplace_component_and_initialize()` function. The action
+ * lists for the `TestGoto` and `TestRepeatUntil` phases are fairly simple and
+ * not the focus of this discussion.
+ *
+ * Having discussed the metavariables and component, let us now look at the test
+ * case.
+ *
+ * \snippet Test_Goto.cpp test case
+ *
+ * Just as for the `TerminatePhase` test we have some type aliases to reduce the
+ * amount of typing needed and to make the test easier to read. The runner is
+ * again created with the default constructor. However, the component is now
+ * inserted using the `ActionTesting::emplace_component_and_initialize()`
+ * function. The third argument to `emplace_component_and_initialize` is a
+ * `tuples::TaggedTuple` of the simple tags, which is to be populated with the
+ * initial values for the component. Note that there is no call to
+ * `set_phase(Metavariables::Phase::Initialization)`: this and the required
+ * action invokation is handled internally by
+ * `ActionTesting::emplace_component_and_initialize()`.
+ *
```

Comment:
does this mean there is no way to do introspection between initialization actions?

Reply:
There is only one initialization action provided by the testing infrastructure. You can test any action in any way you want. The way you would test "initialization actions" is by initializing the runner with whatever input the initialization actions need. For example, the Domain initialization action requires there be the `InitialExtents` in the DataBox. Thus, the initialize part of `emplace_component_and_initialize` sets the `InitialExtents`. That is, it is initializing the test and has nothing to do with initialization actions.